File: go/supersandbox/sandbox.go
package supersandbox

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"time"
)

// SandboxHandle is a rich wrapper around a sandbox with convenience methods.
type SandboxHandle struct {
	ID             string
	Status         SandboxStatus
	Metadata       map[string]string
	Entrypoint     []string
	ExpiresAt      *time.Time
	CreatedAt      time.Time
	LastActivityAt *time.Time

	http *httpClient
}

func sandboxHandleFromInfo(info SandboxInfo, h *httpClient) *SandboxHandle {
	return &SandboxHandle{
		ID:             info.ID,
		Status:         info.Status,
		Metadata:       info.Metadata,
		Entrypoint:     info.Entrypoint,
		ExpiresAt:      info.ExpiresAt,
		CreatedAt:      info.CreatedAt,
		LastActivityAt: info.LastActivityAt,
		http:           h,
	}
}

// Tasks returns the Tasks sub-resource for this sandbox.
func (s *SandboxHandle) Tasks() *Tasks {
	return &Tasks{sandboxID: s.ID, http: s.http}
}

// Delete removes this sandbox.
func (s *SandboxHandle) Delete(ctx context.Context) error {
	return s.http.delete(ctx, fmt.Sprintf("/sandboxes/%s", s.ID))
}

// Pause suspends this sandbox.
func (s *SandboxHandle) Pause(ctx context.Context) error {
	return s.http.post(ctx, fmt.Sprintf("/sandboxes/%s/pause", s.ID), nil, nil)
}

// Resume resumes a paused sandbox.
func (s *SandboxHandle) Resume(ctx context.Context) error {
	return s.http.post(ctx, fmt.Sprintf("/sandboxes/%s/resume", s.ID), nil, nil)
}

// RenewExpiration extends the sandbox TTL.
func (s *SandboxHandle) RenewExpiration(ctx context.Context, ttlSeconds int) (*RenewExpirationResponse, error) {
	body := map[string]int{"timeout": ttlSeconds}
	var resp RenewExpirationResponse
	if err := s.http.put(ctx, fmt.Sprintf("/sandboxes/%s/expiration", s.ID), body, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// UpdateResourceLimits changes the CPU/memory limits for this sandbox.
func (s *SandboxHandle) UpdateResourceLimits(ctx context.Context, limits ResourceLimits) (*UpdateResourceLimitsResponse, error) {
	body := map[string]ResourceLimits{"resourceLimits": limits}
	var resp UpdateResourceLimitsResponse
	if err := s.http.put(ctx, fmt.Sprintf("/sandboxes/%s/resources", s.ID), body, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// UpdateEnv sets environment variables on the sandbox.
func (s *SandboxHandle) UpdateEnv(ctx context.Context, env map[string]string) (*UpdateEnvResponse, error) {
	body := map[string]map[string]string{"env": env}
	var resp UpdateEnvResponse
	if err := s.http.put(ctx, fmt.Sprintf("/sandboxes/%s/env", s.ID), body, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// GetEndpoint returns the public address for a port exposed by the sandbox.
func (s *SandboxHandle) GetEndpoint(ctx context.Context, port int, public bool) (*Endpoint, error) {
	params := url.Values{}
	if public {
		params.Set("public", "true")
	}
	var ep Endpoint
	if err := s.http.get(ctx, fmt.Sprintf("/sandboxes/%s/endpoints/%d", s.ID, port), params, &ep); err != nil {
		return nil, err
	}
	return &ep, nil
}

// GetLogs returns the last tail lines of container logs.
func (s *SandboxHandle) GetLogs(ctx context.Context, tail int) (string, error) {
	params := url.Values{}
	params.Set("tail", strconv.Itoa(tail))
	params.Set("follow", "false")
	var logs string
	if err := s.http.get(ctx, fmt.Sprintf("/sandboxes/%s/logs", s.ID), params, &logs); err != nil {
		return "", err
	}
	return logs, nil
}

// StreamLogs returns a channel that emits log lines in real time.
func (s *SandboxHandle) StreamLogs(ctx context.Context, tail int) <-chan StreamLine {
	params := url.Values{}
	params.Set("tail", strconv.Itoa(tail))
	params.Set("follow", "true")
	return s.http.stream(ctx, fmt.Sprintf("/sandboxes/%s/logs", s.ID), params)
}

// Terminal opens a WebSocket PTY session on the sandbox.
func (s *SandboxHandle) Terminal(ctx context.Context, opts ...TerminalOption) (*TerminalSession, error) {
	if s.Status.State != "Running" {
		return nil, fmt.Errorf("sandbox %s is not running (state: %s)", s.ID, s.Status.State)
	}
	return newTerminalSession(ctx, s.http, s.ID, opts...)
}

func (s *SandboxHandle) String() string {
	return fmt.Sprintf("SandboxHandle(id=%s, state=%s)", s.ID, s.Status.State)
}

File: python/src/supersandbox/sandbox.py
"""Sandbox — high-level wrapper around a running sandbox instance."""

from __future__ import annotations

from contextlib import asynccontextmanager
from datetime import datetime
from typing import TYPE_CHECKING, Any, AsyncIterator, Dict, List, Optional, Union

from ._http import AsyncHTTPClient
from ._utils.errors import intercept_errors
from .exceptions import SandboxNotRunningError
from .models import (
    CreateSandboxResponse,
    Endpoint,
    ImageSpec,
    RenewExpirationResponse,
    Sandbox as _SandboxData,
    SandboxStatus,
    UpdateEnvResponse,
    UpdateResourceLimitsResponse,
)
from .tasks import Tasks
from .terminal import TerminalSession

# Union of the two shapes returned by the API (create vs. get/list)
_AnyData = Union[_SandboxData, CreateSandboxResponse]


class Sandbox:
    """A running sandbox instance.

    Returned by :meth:`AsyncSuperSandbox.create`, :meth:`AsyncSuperSandbox.get`,
    and :meth:`AsyncSuperSandbox.list`.

    Sub-resources are accessed as properties::

        task = await sandbox.tasks.submit("echo hello")
        async with sandbox.terminal() as term:
            await term.send("ls /\\n")
    """

    def __init__(
        self,
        data: _AnyData,
        http: AsyncHTTPClient,
        base_url: str,
        api_key: str,
    ) -> None:
        self.id: str = data.id
        self.image: Optional[ImageSpec] = getattr(data, "image", None)
        self.status: SandboxStatus = data.status
        self.metadata: Optional[Dict[str, str]] = data.metadata
        self.entrypoint: List[str] = data.entrypoint
        self.expires_at: Optional[datetime] = data.expires_at
        self.created_at: datetime = data.created_at
        self.last_activity_at: Optional[datetime] = getattr(data, "last_activity_at", None)

        self._http = http
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key

    # ── Sub-resources ─────────────────────────────────────────────────────────

    @property
    def tasks(self) -> Tasks:
        """Background task execution on this sandbox."""
        return Tasks(self._http, self.id)

    @asynccontextmanager
    @intercept_errors
    async def terminal(self) -> AsyncIterator[TerminalSession]:
        """Open an interactive WebSocket terminal session.

        Raises:
            SandboxNotRunningError: If the sandbox is not in ``Running`` state.

        Usage::

            async with sandbox.terminal() as term:
                await term.send("echo hello\\n")
                print(await term.receive())
        """
        if self.status.state != "Running":
            raise SandboxNotRunningError(
                f"Sandbox {self.id!r} is not running (state={self.status.state!r})"
            )
        token_data = await self._http.post(f"/sandboxes/{self.id}/terminal/token")
        ws_url = token_data.get("url") or (
            self._base_url.replace("https://", "wss://").replace("http://", "ws://")
            + f"/sandboxes/{self.id}/terminal?token={token_data['token']}"
        )
        session = TerminalSession(ws_url, self._api_key)
        async with session as term:
            yield term

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    @intercept_errors
    async def delete(self) -> None:
        """Terminate and delete this sandbox."""
        await self._http.delete(f"/sandboxes/{self.id}")

    @intercept_errors
    async def pause(self) -> None:
        """Pause the sandbox (scales to zero, /workspace is preserved)."""
        await self._http.post(f"/sandboxes/{self.id}/pause")

    @intercept_errors
    async def resume(self) -> None:
        """Resume a paused sandbox."""
        await self._http.post(f"/sandboxes/{self.id}/resume")

    @intercept_errors
    async def renew_expiration(self, expires_at: datetime) -> RenewExpirationResponse:
        """Extend the sandbox TTL to ``expires_at``."""
        data = await self._http.post(
            f"/sandboxes/{self.id}/renew-expiration",
            json={"expiresAt": expires_at.isoformat()},
        )
        return RenewExpirationResponse.model_validate(data)

    @intercept_errors
    async def update_resource_limits(
        self,
        *,
        cpu: Optional[str] = None,
        memory: Optional[str] = None,
        storage: Optional[str] = None,
    ) -> UpdateResourceLimitsResponse:
        """Update CPU / memory / storage limits while the sandbox is running."""
        limits: Dict[str, str] = {}
        if cpu:
            limits["cpu"] = cpu
        if memory:
            limits["memory"] = memory
        if storage:
            limits["storage"] = storage
        data = await self._http.patch(
            f"/sandboxes/{self.id}", json={"resourceLimits": limits}
        )
        return UpdateResourceLimitsResponse.model_validate(data)

    @intercept_errors
    async def update_env(self, env: Dict[str, Optional[str]]) -> UpdateEnvResponse:
        """Replace all user-defined environment variables.

        Pass ``None`` as a value to delete a variable.
        Internal variables (e.g. ``EXECD_TOKEN``) are preserved automatically.
        """
        data = await self._http.put(f"/sandboxes/{self.id}/env", json={"env": env})
        return UpdateEnvResponse.model_validate(data)

    # ── Observability ─────────────────────────────────────────────────────────

    @intercept_errors
    async def get_endpoint(
        self, port: int, *, use_server_proxy: bool = False
    ) -> Endpoint:
        """Get the public endpoint for a port exposed by the sandbox.

        Args:
            port:             Port number inside the sandbox.
            use_server_proxy: Route through the server-side HTTP proxy.
        """
        params: Dict[str, Any] = {}
        if use_server_proxy:
            params["use_server_proxy"] = "true"
        data = await self._http.get(
            f"/sandboxes/{self.id}/endpoints/{port}", params=params or None
        )
        return Endpoint.model_validate(data)

    @intercept_errors
    async def get_logs(self, *, tail: int = 100) -> str:
        """Return the last ``tail`` lines of container logs as a string."""
        params: Dict[str, Any] = {"tail": tail, "follow": False}
        return await self._http.get(f"/sandboxes/{self.id}/logs", params=params)

    async def stream_logs(self, *, tail: int = 100) -> AsyncIterator[str]:
        """Async generator that streams container log lines in real time."""
        params: Dict[str, Any] = {"tail": tail, "follow": True}
        async for line in self._http.stream_get(f"/sandboxes/{self.id}/logs", params=params):
            yield line

    def __repr__(self) -> str:
        return f"Sandbox(id={self.id!r}, state={self.status.state!r})"
